### PR TITLE
Fix #976: Break claims value in array into multiple claims

### LINKED ISSUE #976: OIDC handler bug in user info response handling for multiple claims of same type
Imagine user info endpoint returns this:

```
{"sub":"88421113","email":"BobSmith@email.com","email_verified":true,"role":["role1","role2"]}
```

This line: https://github.com/aspnet/Security/blob/dev/src/Microsoft.AspNetCore.Authentication.OpenIdConnect/OpenIdConnectHandler.cs#L788

Adds the role claim with this as the value: `[ "role1", "role2" ]`. 

It should break the array into multiple claims. Here's an extension method that already does this: https://github.com/IdentityModel/IdentityModel2/blob/dev/src/IdentityModel/Client/JObjectExtensions.cs#L12

### PR DESCRIPTION
**Pull request by troydai**
#976

### CODE REVIEW

**Comment by PinpointTownes** [comment]
@troydai curious: have you considered using a smarter approach like the one used by Wilson? It has a few benefits: it "preserves" the claim value type and uses a specific value type when it can't flatten the JSON payload.

**Comment by troydai** [comment]
Yes.

Wilson's approach looks comprehensive, but I haven't fully evaluated it yet. The reason why I hesitate to adopt here is that the change to be brought in will be much larger than it is needed to address the target issue which for now involves array only. I tend to keep the change smaller at this stage. 

In addition, I'm wondering if completely change how we break down the claims may break customers, please let me know if it something from your perspective worth worries about.

**Comment by Tratcher** [comment]
(linked for my convenience) https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/master/src/System.IdentityModel.Tokens.Jwt/JwtPayload.cs#L307

@PinpointTownes do you have scenarios for any of the data types covered by wilson right now? They look very generic and it's not clear how they map back to JSON structures. E.g. `IEnumberable<object>` or `IDictionary<string, object>`

**Comment by Tratcher** [comment]
Or were you referring to this part? https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/master/src/System.IdentityModel.Tokens.Jwt/JwtPayload.cs#L392-L399

**Comment by PinpointTownes** [comment]
> In addition, I'm wondering if completely change how we break down the claims may break customers, please let me know if it something from your perspective worth worries about.

Well, this change itself is already a breaking change for people who rely on the fact the OIDC middleware doesn't properly format "array claims". Preserving the claim value type wouldn't break them harder :smile: 

@Tratcher Wilson's approach is a bit complex because `JwtPayload` is a generic dictionary that accepts both primitives types (like `string`, `int` or enumerations of primitive types) and JSON.NET types (i.e types derived from `JToken`). I suppose this approach was adopted for performance, to avoid instantiating `JToken` objects when you only want to use primitive types.

Here, we always deal with `JObject`, so the related code will be much simpler. The code snippet you mentioned in your last post + https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/master/src/System.IdentityModel.Tokens.Jwt/JwtPayload.cs#L547-L582 should be enough to preserve the claim value type and to assign a JSON-specific type when we can't flatten it.

**Comment by troydai** [comment]
How about address the array claims issue in the original issue first and then follow up a more comprehensive solution to the overall situation? I believe we need a little bit more time to evaluate and test the existing solution.

**Comment by Tratcher** [comment]
@PinpointTownes it looks like those are additional issues beyond flattening arrays to multiple claims. I'd want to see a representative user-info dataset that took advantage of these different structures.

**Comment by PinpointTownes** [comment]
@Tratcher I'm not suggesting updating the code used to flatten the claims. I'd just like the claim value types to be preserved (which is completely orthogonal).

**Comment by Tratcher** [comment]
Yes, file a different bug and we'll look into it.

File: src/Microsoft.AspNetCore.Authentication.OpenIdConnect/Utility/ClaimsHelper.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+                }
+                else
+                {
+                    AddClaimsToIdentity(pair.Value, identity, pair.Key, issuer);
+                }
+            }
+        }
+
+        private static void AddClaimsToIdentity(JToken item, ClaimsIdentity identity, string key, string issuer)
+            => identity.AddClaim(new Claim(key, item.ToString(), ClaimValueTypes.String, issuer));
```
Comment (PinpointTownes) [inline src/Microsoft.AspNetCore.Authentication.OpenIdConnect/Utility/ClaimsHelper.cs]: FWIW, the existing code is most likely invalid, as `Claim`'s constructor always throws an exception for null values (but not for empty values).

**Comment by PinpointTownes** [comment]
Done: https://github.com/aspnet/Security/issues/998

File: src/Microsoft.AspNetCore.Authentication.OpenIdConnect/Utility/ClaimsHelper.cs (hunk 1 above)
Comment (troydai) [inline src/Microsoft.AspNetCore.Authentication.OpenIdConnect/Utility/ClaimsHelper.cs]: Addressed in next commit. If the value is null, turns it into string.Empty.

**Comment by troydai** [comment]
The PR is approved, I'll merge once the CI passed.

**Pull request merged**